File: ruby/ratio.rb
class Ratio
    attr_reader :d,:n
    def initialize(n,d)
        if d==0
            @n=1
            @d=0
        else
            g=n.abs.gcd(d.abs)
            g = d < 0 ? -g : g
            @n=n/g
            @d=d/g
        end

    end

    def to_str()
        if @d==0
            "NaN"
        else
            if @d==1
                @n.to_s
            else
                @n.to_s+"/"+@d.to_s
            end
        end
    end

    def +(other)
        if @d!=0&&other.d!=0
            Ratio.new(@n * other.d + other.n * @d, @d * other.d)
        else
            Ratio.new(1,0)
        end
    end

    def -(other)
        if @d!=0&&other.d!=0
            Ratio.new(@n * other.d - other.n * @d, @d * other.d)
        else
            Ratio.new(1,0)
        end
    end

    def *(other)
        if @d!=0&&other.d!=0
            Ratio.new(@n*other.n, @d * other.d)
        else
            Ratio.new(1,0)
        end
    end

    def /(other)
        if @d!=0&&other.d!=0
            Ratio.new(@n*other.d, @d * other.n)
        else
            Ratio.new(1,0)
        end
    end
end

File: haskell/app/Ratio.hs
module Ratio where

import qualified Data.Ratio                    as R

data Ratio= Ratio Int Int|NaN deriving (Eq)

ratio :: Int -> Int -> Ratio
ratio _ 0 = NaN
ratio n d = Ratio (n `div` g) (d `div` g)
    where g = signum d * gcd (abs n) (abs d)

instance Show Ratio where
    show NaN="NaN"
    show (Ratio n 1)=show n
    show (Ratio n d)=(show n)++"/"++(show d)

instance Num Ratio where
    NaN + _ = NaN
    _ + NaN = NaN
    (Ratio an ad) + (Ratio bn bd)=ratio (an * bd + bn * ad) (ad*bd)

    NaN * _ = NaN
    _ * NaN = NaN
    (Ratio an ad) * (Ratio bn bd)=ratio (an * bn) (ad*bd)

    negate NaN=NaN
    negate (Ratio n d) = ratio (negate n) d

    abs NaN=NaN
    abs (Ratio n d)=ratio (abs n) d

    fromInteger x=ratio (fromInteger x) 1

    signum NaN=NaN
    signum (Ratio n _)=ratio (signum n) 1


instance Fractional Ratio where
    recip NaN=NaN
    recip (Ratio n d)=ratio d n

    fromRational x=ratio ((fromInteger . R.numerator) x) ((fromInteger . R.denominator) x)
